crates/milli/src/vector/extractor.rs

```diff
@@ -0,0 +1,219 @@
+    let new_input = new_input?;
+    Ok(match (old_input, new_input) {
+        (Some(old), Some(new)) if old == new => ExtractorDiff::Unchanged,
+        (None, None) => ExtractorDiff::Unchanged,
+        (None, Some(input)) => ExtractorDiff::Added(input),
+        (Some(_), None) => ExtractorDiff::Removed,
+        (Some(_), Some(input)) => ExtractorDiff::Updated(input),
+    })
+}
+
+pub enum ExtractorDiff<Input> {
+    Removed,
+    Added(Input),
+    Updated(Input),
+    Unchanged,
+}
+
+impl<Input> ExtractorDiff<Input> {
+    pub fn into_input(self) -> Option<Input> {
+        match self {
+            ExtractorDiff::Removed => None,
+            ExtractorDiff::Added(input) => Some(input),
+            ExtractorDiff::Updated(input) => Some(input),
+            ExtractorDiff::Unchanged => None,
+        }
+    }
+
+    pub fn needs_change(&self) -> bool {
+        match self {
+            ExtractorDiff::Removed => true,
+            ExtractorDiff::Added(_) => true,
+            ExtractorDiff::Updated(_) => true,
+            ExtractorDiff::Unchanged => false,
+        }
+    }
+
+    pub fn into_list_of_changes(
+        named_diffs: impl IntoIterator<Item = (String, Self)>,
+    ) -> BTreeMap<String, Option<Input>> {
+        named_diffs
+            .into_iter()
+            .filter(|(_, diff)| diff.needs_change())
+            .map(|(name, diff)| (name, diff.into_input()))
+            .collect()
+    }
+}
+
+pub struct DocumentTemplateExtractor<'a, 'b> {
+    template: &'a Prompt,
+    doc_alloc: &'a Bump,
+    field_id_map: &'a RefCell<GlobalFieldsIdsMap<'b>>,
+}
+
+impl<'a, 'b> DocumentTemplateExtractor<'a, 'b> {
+    pub fn new(
+        template: &'a Prompt,
+        doc_alloc: &'a Bump,
+        field_id_map: &'a RefCell<GlobalFieldsIdsMap<'b>>,
+    ) -> Self {
+        Self { template, doc_alloc, field_id_map }
+    }
+}
+
+impl<'doc> Extractor<'doc> for DocumentTemplateExtractor<'doc, '_> {
+    type DocumentMetadata = &'doc str;
+    type Input = &'doc str;
+    type Error = RenderPromptError;
+
+    fn extractor_id(&self) -> u8 {
+        0
+    }
+
+    fn extract<'a, D: Document<'a> + Debug>(
+        &self,
+        doc: D,
+        external_docid: &Self::DocumentMetadata,
+    ) -> Result<Option<Self::Input>, Self::Error>
+    where
+        'doc: 'a,
+    {
+        Ok(Some(self.template.render_document(
+            external_docid,
+            doc,
+            self.field_id_map,
+            self.doc_alloc,
+        )?))
+    }
+}
+
+pub struct RequestFragmentExtractor<'a> {
+    fragment: &'a JsonTemplate,
+    extractor_id: u8,
+    doc_alloc: &'a Bump,
+}
+
+impl<'a> RequestFragmentExtractor<'a> {
+    pub fn new(fragment: &'a RuntimeFragment, doc_alloc: &'a Bump) -> Self {
+        Self { fragment: &fragment.template, extractor_id: fragment.id, doc_alloc }
+    }
+}
+
+impl<'doc> Extractor<'doc> for RequestFragmentExtractor<'doc> {
+    type DocumentMetadata = ();
+    type Input = Value;
```

Comment:
consider renaming `Input`?

Reply:
Did not rename, but added documentation clarifying that we are talking about an "input to an embedder"